src/chunking.ts

```diff
@@ -0,0 +1,180 @@
+/* eslint-disable no-bitwise, no-plusplus, no-mixed-operators */
+
+import type { Json } from '@metamask/utils';
+import { type TransportChunkFrame } from './types';
+
+const INT32_MAX = 0x7fffffff;
+
+let _randomCounter = 0 | 0;
+
+/**
+ * Creates a random ID for chunked messages. Always <= Int32 max value.
+ *
+ * @returns a random number that can be used as an ID
+ */
+function getNextId(): number {
+  // increment the counter and wrap around if it exceeds Int32 max value
+  _randomCounter = (_randomCounter + 1) & INT32_MAX;
+  return _randomCounter;
+}
+
+// half a "frame" (at 60 fps) per chunk, in milliseconds
+const FRAME_BUDGET_MS = 1000 / 60 / 2;
+
+/**
+ * Yield control to allow other tasks to run. Uses scheduler.yield() if available
+ * (modern browsers), otherwise falls back to setTimeout (older browsers).
+ */
+const tick = (() => {
+  return globalThis.scheduler?.yield
+    ? globalThis.scheduler.yield.bind(globalThis.scheduler)
+    : async () => await new Promise<void>((resolve) => setTimeout(resolve, 0));
+})();
+
+/**
+ * Compute the UTF-8 byte length for a single code point
+ *
+ * Uses code point ranges instead of `new TextEncoder().encode("x").byteLength`
+ * for performance reasons.
+ *
+ * @param cp - the code point
+ * @returns the UTF-8 byte length
+ */
+function utf8Len(cp: number): number {
+  if (cp <= 0x7f) {
+    // e.g., new TextEncoder().encode("x").byteLength === 1
+    return 1;
+  }
+  if (cp <= 0x7ff) {
+    // e.g., new TextEncoder().encode("¢").byteLength === 2
+    return 2;
+  }
+  if (cp <= 0xffff) {
+    // e.g., new TextEncoder().encode("✓").byteLength === 3
+    return 3;
+  }
+  // e.g., new TextEncoder().encode("😀").byteLength === 4
+  return 4;
+}
+
+/**
+ * Finds the max cut index such that:
+ * `bytes(JSON.stringify(s.slice(start, cut))) <= maxLength`
+ *
+ * We do this by starting with `used = 2`, then we add per-code-point costs for
+ * the slice until we’d exceed maxLength.
+ * @param json - the full JSON string
+ * @param start - the starting index for the slice
+ * @param maxBytes - the maximum allowed length in _bytes_
+ */
+function getJsonSliceIndex(json: string, start: number, maxBytes: number) {
+  // Start with the overhead bytes that JSON.stringify will add (2 bytes for
+  // the outer quotation marks)
+  let bytesUsed = 2;
+
+  let i = start;
+  while (i < json.length) {
```

Comment:
Forgive me for asking the obvious question, do we have any figures for the performance gain to warrant the readability / complexity hit versus `new TextEncoder().encode` you noted above.

For our standard use cases, will this make a meaningful difference to the performance perceived by the user?

Naturally I'm concerned that we can loop manually, character by character in JS, and it be faster than what the standard `TextEncoder` library is doing?